x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,182 @@
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	ack exported.Acknowledgement,
+) exported.Acknowledgement {
+	logger := k.Logger(ctx)
+
+	params := k.GetParams(ctx)
+	claimsParams := k.claimsKeeper.GetParams(ctx)
+
+	// check channels from this chain (i.e destination)
+	if !params.EnableWithdraw ||
+		!claimsParams.IsAuthorizedChannel(packet.DestinationChannel) ||
+		claimsParams.IsEVMChannel(packet.DestinationChannel) {
+		// return original ACK if:
+		// - withdraw is disabled globally
+		// - channel is not authorized
+		// - channel is an EVM channel
+		return ack
+	}
+
+	// unmarshal packet data to obtain the sender and recipient
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	// case 1: sender ≠ recipient.
+	// Withdraw is only possible for addresses in which the sender = recipient.
+	// Continue to the next IBC middleware by returning the original ACK.
+	if !sender.Equals(recipient) {
+		return ack
+	}
+
+	// get the recipient account
+	account := k.accountKeeper.GetAccount(ctx, recipient)
+
+	// Case 2. recipient pubkey is a supported key (eth_secp256k1, amino multisig, ed25519)
+	// ==> Continue and return success ACK as the funds are not stuck on chain
+	if account != nil &&
+		account.GetPubKey() != nil &&
+		evmos.IsSupportedKey(account.GetPubKey()) {
+		return ack
+	}
+
+	// NOTE: Since destination channel is authorized and not from an EVM chain, we know that
+	// only secp256k1 keys are supported in the source chain. This means that we can now
+	// initiate the withdraw logic
+
+	// transfer the balance back to the sender address
+	destPort := packet.DestinationPort
+	destChannel := packet.DestinationChannel
+	balances := sdk.Coins{}
+
+	k.bankKeeper.IterateAccountBalances(ctx, recipient, func(coin sdk.Coin) (stop bool) {
```

Comment:
just a sanity check. Does this account for tokens in vesting? Is it possible for there to be tokens in vesting for a secp address?

Reply:
>Does this account for tokens in vesting?

it doesn't account for vesting tokens, we can prob add a check to ignore the withdrawal for vesting accounts

>Is it possible for there to be tokens in vesting for a secp address?

no, the secp256k1 vesting accounts would have needed to be generated at genesis.